includes/data-port/models/class-sensei-data-port-model.php

```diff
@@ -0,0 +1,260 @@
+<?php
+/**
+ * File containing the Sensei_Data_Port_Model class.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Defines the expected data to port to/from and handles the port.
+ */
+abstract class Sensei_Data_Port_Model {
+	/**
+	 * Data in its array form.
+	 *
+	 * @var array
+	 */
+	private $data;
+
+	/**
+	 * Post ID of top-most post. This will be null if creating a new post.
+	 *
+	 * @var int
+	 */
+	private $post_id;
+
+	/**
+	 * Sensei_Data_Port_Model constructor.
+	 */
+	private function __construct() {
+		// Silence is golden.
+	}
+
+	/**
+	 * Set up item from an array.
+	 *
+	 * @param array $data Data to restore item from.
+	 *
+	 * @return static
+	 */
+	public static function from_source_array( $data ) {
+		$self = new static();
+		$self->restore_from_source_array( $data );
+
+		if ( $self->is_valid() ) {
+			$post_id = $self->get_existing_post_id();
+			if ( $post_id ) {
+				$self->set_post_id( $post_id );
+			}
+		}
+
+		return $self;
+	}
+
+	/**
+	 * Check to see if the post already exists in the database.
+	 *
+	 * @return int
+	 */
+	abstract protected function get_existing_post_id();
+
+	/**
+	 * Create a new post or update an existing post.
+	 *
+	 * @return true|WP_Error
+	 */
+	abstract public function sync_post();
+
+	/**
+	 * Check if all required fields are set.
+	 *
+	 * @return bool
+	 */
+	public function is_valid() {
+		$data = $this->get_data();
+
+		foreach ( static::get_schema() as $field => $field_config ) {
+			if ( isset( $data[ $field ] ) ) {
+				continue;
+			}
+
+			// If the field is required, it must be set.
+			if ( ! empty( $field_config['required'] ) ) {
+				return false;
+			}
+
+			// If a default exists as well as a pattern, a `null` value is for a field that didn't match the pattern.
+			if (
+				array_key_exists( $field, $data )
+				&& ! empty( $field_config['default'] )
+				&& ! empty( $field_config['pattern'] )
+			) {
```

Comment:
So if a user wants to delete a value for a field, it will have an empty value for the specific column. For this case if there is no default the actual value is going to be null. How are we going to differentiate between this and the case where the regexp failed? Do we need to document that if regexp is used then default should be defined too?

Reply:
Talked on Slack; trying out empty string as default to differentiate 99e500b428b24d1c9ce450eae614d7aee1116c68